pennylane_lightning/src/simulator/DynamicDispatcher.cpp

```diff
@@ -0,0 +1,212 @@
+// Copyright 2022 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+/**
+ * @file DynamicDispatcher.cpp
+ * Register all gate and generator implementations
+ */
+#include "DynamicDispatcher.hpp"
+#include "AvailableKernels.hpp"
+#include "Constant.hpp"
+#include "OpToMemberFuncPtr.hpp"
+#include "SelectGateOps.hpp"
+#include "SimulatorUtil.hpp"
+
+using namespace Pennylane;
+
+namespace {
+/**
+ * @brief return a lambda function for the given kernel and gate operation
+ *
+ * As we want the lamba function to be stateless, kernel and gate_op are
+ * template paramters (or the functions can be consteval in C++20).
+ * In C++20, one also may use a template lambda function instead.
+ *
+ * @tparam PrecisionT Floating point precision of underlying statevector data.
+ * @tparam ParamT Floating point type for parameters>
+ * @tparam GateImplementation Gate implementation class.
+ * @tparam gate_op Gate operation to make a functor.
+ */
+template <class PrecisionT, class ParamT, class GateImplementation,
+          GateOperation gate_op>
+constexpr auto gateOpToFunctor() {
+    return [](std::complex<PrecisionT> *data, size_t num_qubits,
+              const std::vector<size_t> &wires, bool inverse,
+              const std::vector<PrecisionT> &params) {
+        constexpr auto func_ptr =
+            GateOpToMemberFuncPtr<PrecisionT, ParamT, GateImplementation,
+                                  gate_op>::value;
+        assert(params.size() ==
+               static_lookup<gate_op>(Constant::gate_num_params));
+        callGateOps(func_ptr, data, num_qubits, wires, inverse, params);
+    };
+}
+
+/// @cond DEV
+/**
+ * @brief Internal recursion function for constructGateOpsFunctorTuple
+ *
+ * @return Tuple of gate operations and corresponding GateImplementation member
+ * function pointers.
+ */
+template <class PrecisionT, class ParamT, class GateImplementation,
```

Comment:
This is beautiful compile-time magic :magic_wand:

Reply:
Thanks! :heart: